### PR TITLE
Migration and validation APIs for version upgrade

### PR DESCRIPTION
**Pull request by hardik-k-shah**
*Issue #, if available:*

*Description of changes:*
Migration and valdiation APIs for version upgrade 

Cherry-pick of various commits: 
https://github.com/opendistro-for-elasticsearch/security/commit/0286850f413016b1586a13189a9bb9d9b05a9bd3
https://github.com/opendistro-for-elasticsearch/security/commit/472d652261c3a19affecfa594b93f7052dc20a68
https://github.com/opendistro-for-elasticsearch/security/commit/2259e6b743f4539e1798ef625d93d2972152a4fa
https://github.com/opendistro-for-elasticsearch/deprecated-security-advanced-modules/commit/8cf60f21e650aa16a77c39339a2455e7603c130b

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by debjanibnrj** [comment]
nit: change the commit title from valdiation to validation

**Comment by hardik-k-shah** [comment]
> nit: change the commit title from valdiation to validation

done

File: src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/ConfigurationRepository.java [hunk 1]
```diff
@@ -101,6 +102,7 @@ private ConfigurationRepository(Settings settings, final Path configPath, Thread
         this.clusterService = clusterService;
         this.auditLog = auditLog;
         this.configurationChangedListener = new ArrayList<>();
+        this.acceptInvalid = settings.getAsBoolean(ConfigConstants.OPENDISTRO_SECURITY_UNSUPPORTED_ACCEPT_INVALID_CONFIG, false);
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/ConfigurationRepository.java:105]: Curious why we have this property - Whats the use case of starting up nodes always accepting invalid config? 
If it's just for loading during migration API call, we can just overload the function to do it?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/MigrateApiAction.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            badRequestResponse(channel, "Can not migrate configuration because it was already migrated.");
+            return;
+        }
+
+        final SecurityDynamicConfiguration<ConfigV6> configV6 = (SecurityDynamicConfiguration<ConfigV6>) loadedConfig;
+        final SecurityDynamicConfiguration<ActionGroupsV6> actionGroupsV6 = (SecurityDynamicConfiguration<ActionGroupsV6>) load(CType.ACTIONGROUPS, true);
+        final SecurityDynamicConfiguration<InternalUserV6> internalUsersV6 = (SecurityDynamicConfiguration<InternalUserV6>) load(CType.INTERNALUSERS, true);
+        final SecurityDynamicConfiguration<RoleV6> rolesV6 = (SecurityDynamicConfiguration<RoleV6>) load(CType.ROLES, true);
+        final SecurityDynamicConfiguration<RoleMappingsV6> rolesmappingV6 = (SecurityDynamicConfiguration<RoleMappingsV6>) load(CType.ROLESMAPPING, true);
+
```
Comment (sujithvm) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/MigrateApiAction.java:110]: Need to add NodesDn also to the migration

File: src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/ConfigurationRepository.java (hunk 1 above)
Comment (hardik-k-shah) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/configuration/ConfigurationRepository.java:105]: accepting invalid config is configuration parameter you can set in elasticsearch.yml 

You can configure if you want to accept invalid/non-supported 6.x configuration and still migrate remaining valid configuration to 7.x format and still start node. 

7.x can still read 6.x configuration and work without any problem. 

If you migrate configuration with accept invalid configuration to true, you need that flag everytime when node boots up to ignore invalid configuration present in index.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/MigrateApiAction.java (hunk 2 above)
Comment (hardik-k-shah) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/MigrateApiAction.java:112]: added.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/NonValidatingObjectMapper.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import org.elasticsearch.SpecialPermission;
+
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.type.TypeFactory;
+
+public class NonValidatingObjectMapper {
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/NonValidatingObjectMapper.java:47]: The class is used only within `SecurityDynamicConfiguration`, can it be made `private static` inside `SecurityDynamicConfiguration`?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/NonValidatingObjectMapper.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    private static final ObjectMapper nonValidatingObjectMapper = new ObjectMapper();
```
Comment (vrozov) [inline src/main/java/com/amazon/opendistroforelasticsearch/security/NonValidatingObjectMapper.java:48]: Is it necessary to make `nonValidatingObjectMapper` a singleton?

File: src/main/java/com/amazon/opendistroforelasticsearch/security/NonValidatingObjectMapper.java (hunk 3 above)
Comment (hardik-k-shah) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/NonValidatingObjectMapper.java:47]: these is mostly cherry pick from 7.1 branch which we missed merging to master branch. 
I would prefer to make separate PR for suggested refactoring if any.

File: src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/MigrateApiAction.java (hunk 2 above)
Comment (sujithvm) [reply src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/MigrateApiAction.java:112]: Need to add to validate api too ?

**Pull request merged**